### PR TITLE
[libdwarf] New port

### PR DESCRIPTION
**Pull request by jeremy-rifkin**
Trying to add libdwarf. Needed for #34217.

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by jeremy-rifkin** [comment]
I'm running into `/usr/bin/ld: cannot find -lZLIB` locally. Not sure what I'm doing wrong here.

**Comment by jeremy-rifkin** [comment]
Ok, looks like it's just z not ZLIB for linking.

**Comment by jeremy-rifkin** [comment]
I'm seeing `dwarf_init_finish.c(57): fatal error C1083: Cannot open include file: 'zlib.h': No such file or directory` on windows, not really sure why.

File: ports/libdwarf/v0.8.0-patches.patch [hunk 1]
```diff
@@ -0,0 +1,165 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 28076ee4..4ebc3a4c 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -193,77 +193,80 @@ check_c_source_compiles([=[
+   }]=] HAVE_STDAFX_H)
+ #message(STATUS "Checking have windows stdafx.h... ${HAVE_STDAFX_H}")
+ 
+-set(CMAKE_REQUIRED_LIBRARIES z)
```
Comment (dg0yt) [inline ports/libdwarf/v0.8.0-patches.patch:9]: Maybe just
```suggestion
 find_package(ZLIB REQUIRED)
 add_library(z ALIAS ZLIB::ZLIB)
 set(CMAKE_REQUIRED_LIBRARIES z)
```

File: ports/libdwarf/v0.8.0-patches.patch [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+-      }
+-      return 0;
+-  } ]=]  HAVE_ZLIB_H )
+-set(CMAKE_REQUIRED_LIBRARIES)
+-if (HAVE_ZLIB) 
+-  # For linking in libz
+-  set(DW_FZLIB "z")
+-endif()
+-
+-set(CMAKE_REQUIRED_LIBRARIES zstd )
```
Comment (dg0yt) [inline ports/libdwarf/v0.8.0-patches.patch:44]:
```suggestion
 find_package(zstd CONFIG REQUIRED)
 if(TARGET zstd::libzstd_shared)
   add_library(zstd ALIAS zstd::libzstd_shared)
 else()
   add_library(zstd ALIAS zstd::libzstd_static)
 endif()
```

**Comment by dg0yt** [comment]
You must transfer these suggestions by hand.
Baseline: Use CMake targets to carry the right library names (z/zlib/zlibd) and include dirs (vcpkg_root/include). Use alias targets to keep the patch minimal.

File: ports/libdwarf/v0.8.0-patches.patch (hunk 1 above)
Comment (jeremy-rifkin) [reply ports/libdwarf/v0.8.0-patches.patch:9]: `add_library(z ALIAS ZLIB::ZLIB)` and linking with `z` seems to have helped, thanks!

File: ports/libdwarf/v0.8.0-patches.patch (hunk 2 above)
Comment (jeremy-rifkin) [reply ports/libdwarf/v0.8.0-patches.patch:44]: I've disabled all zstd stuff since libdwarf can get zlib through vcpkg

**Comment by jeremy-rifkin** [comment]
I'm not sure why dlls are being created in static

**Comment by jeremy-rifkin** [comment]
Looked a little more but I'm not really sure what's up here. Unless `VCPKG_LIBRARY_LINKAGE STREQUAL "static"` doesn't do what I think it does. `BUILD_NON_SHARED` feels backwards to me but the current setting is working on everything but the two targets that are failing.

**Comment by jeremy-rifkin** [comment]
Hi @dg0yt and @Adela0814, I think this is now hopefully good to go

File: ports/libdwarf/v0.8.0-patches.patch [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
++#       if (res == destlen) {
++#            /* ALL IS WELL */
++#       }
++#       return 0;
++#   } ]=]  HAVE_ZSTD_H )
++# set(CMAKE_REQUIRED_LIBRARIES)
++# if (HAVE_ZSTD)
++#   # For linking in libzstd
++#   set(DW_FZSTD "zstd")
++# endif()
```
Comment (Adela0814) [inline ports/libdwarf/v0.8.0-patches.patch:148]: To reduce the length of the patch, please use `if (0)` and `endif()` to disable those lines.

**Review by Adela0814 — CHANGES_REQUESTED** [review]

**Comment by Adela0814** [comment]
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

File: ports/libdwarf/v0.8.0-patches.patch (hunk 3 above)
Comment (jeremy-rifkin) [reply ports/libdwarf/v0.8.0-patches.patch:148]: Thanks, that's much better

**Comment by jeremy-rifkin** [comment]
Should be good now @Adela0814, thank you for the review!

File: ports/libdwarf/v0.8.0-patches.patch (hunk 3 above)
Comment (dg0yt) [reply ports/libdwarf/v0.8.0-patches.patch]: Well, could we simply pass the check results as cmake options...

**Comment by Adela0814** [comment]
@jeremy-rifkin  Usage test failed on x64-windows.
Usage:
```
  # this is heuristically generated, and may not be correct
  find_package(libdwarf CONFIG REQUIRED)
  target_link_libraries(main PRIVATE libdwarf::dwarf-static)
```
Error message:
```
1> CMake generation started for configuration: 'x64-Debug'.
1> Command line: "C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="E:\CMakeTest\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE:STRING="E:\\5\vcpkg\\scripts\\buildsystems\\vcpkg.cmake" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" "E:\CMakeTest" 2>&1"
1> Working directory: E:\CMakeTest\out\build\x64-Debug
1> [CMake] -- The C compiler identification is MSVC 19.37.32824.0
1> [CMake] -- The CXX compiler identification is MSVC 19.37.32824.0
1> [CMake] -- Detecting C compiler ABI info
1> [CMake] -- Detecting C compiler ABI info - done
1> [CMake] -- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting C compile features
1> [CMake] -- Detecting C compile features - done
1> [CMake] -- Detecting CXX compiler ABI info
1> [CMake] -- Detecting CXX compiler ABI info - done
1> [CMake] -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting CXX compile features
1> [CMake] -- Detecting CXX compile features - done
1> [CMake] -- Configuring done (6.2s)
1> [CMake] CMake Error at E:/5/vcpkg/installed/x64-windows/share/libdwarf/libdwarf-targets.cmake:60 (set_target_properties):
1> [CMake]   The link interface of target "libdwarf::dwarf-static" contains:
1> [CMake] 
1> [CMake]     ZLIB::ZLIB
1> [CMake] 
1> [CMake]   but the target was not found.  Possible reasons include:
1> [CMake] 
1> [CMake]     * There is a typo in the target name.
1> [CMake]     * A find_package call is missing for an IMPORTED target.
1> [CMake]     * An ALIAS target is missing.
1> [CMake] 
1> [CMake] Call Stack (most recent call first):
1> [CMake]   E:/5/vcpkg/installed/x64-windows/share/libdwarf/libdwarf-config.cmake:2 (include)
1> [CMake]   E:/5/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
1> [CMake]   CMakeTest/CMakeLists.txt:17 (find_package)
1> [CMake] -- Generating done (0.0s)
1> [CMake] CMake Generate step failed.  Build files cannot be regenerated correctly.
1> 'C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="E:\CMakeTest\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE:STRING="E:\\5\vcpkg\\scripts\\buildsystems\\vcpkg.cmake" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" "E:\CMakeTest" 2>&1"' execution failed with error: ''C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="E:\CMakeTest\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE:STRING="E:\\5\vcpkg\\scripts\\buildsystems\\vcpkg.cmake" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" "E:\CMakeTest" 2>&1"' returned with exit code: 1'.
```

**Comment by jeremy-rifkin** [comment]
Thank you, I think the best solution is just to provide a usage text including instructions to `find_package(ZLIB CONFIG REQUIRED)`. It's a bit misleading that the libdwarf target is `dwarf-static` regardless of static or dynamic linkage but I'm not sure it's worth the effort to fix.

 Should be good to review again now.

**Comment by Adela0814** [comment]
> Thank you, I think the best solution is just to provide a usage text including instructions to `find_package(ZLIB CONFIG REQUIRED)`. It's a bit misleading that the libdwarf target is `dwarf-static` regardless of static or dynamic linkage but I'm not sure it's worth the effort to fix.
> 
> Should be good to review again now.

It's better to add `find_dependency(ZLIB)` to `***Config.cmake`.

**Comment by dg0yt** [comment]
@jeremy-rifkin I will send you a PR instead of review comments.

**Comment by dg0yt** [comment]
https://github.com/jeremy-rifkin/vcpkg/pull/1
- Updates dependency import and export. Skips the try-compile for zlib and zstd because it doesn't work well with targets, and we know the result.
- Install proper pkg-config.
- Fixes library linkage, and ensures dllexport.
- Installs the dwarfdump tool.

**Comment by jeremy-rifkin** [comment]
I think it should now be good to go for another review. Thanks so much for all the help!

**Comment by jeremy-rifkin** [comment]
The target is named differently for shared/static, which seems sub-ideal. I might be able to add an alias to take care of that if needed.

**Comment by dg0yt** [comment]
> The target is named differently for shared/static, which seems sub-ideal. I might be able to add an alias to take care of that if needed.

Please restore the usage file. It is the correct solution to the target name variation.

**Review by Adela0814 — DISMISSED** [review]

**Comment by Adela0814** [comment]
No feature needs to test.
Usage tested pass on `x64-windows`.

**Comment by jeremy-rifkin** [comment]
@dg0yt My apologies, I didn't realize you had updated the usage file :)

**Comment by jeremy-rifkin** [comment]
Your usage file works for me locally for x64 static and shared

**Pull request merged**